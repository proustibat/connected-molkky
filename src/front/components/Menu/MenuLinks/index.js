import React, { useState } from 'react';
import { simulateThrow, startFakeSkittles, stopFakeSkittles } from '@root/front/services/api';
import Button from '@components/Button';
import PropTypes from 'prop-types';

const MenuLinks = ({ id, className, style }) => {
  const [isFakeServerRunning, setIsFakerServerRunning] = useState(false);

  const onFakeServerLink = async () => {
    const result = await (isFakeServerRunning ? stopFakeSkittles() : startFakeSkittles());
    if (result) {
      setIsFakerServerRunning(result.mock.skittles);
    }
  };

  return (
    <ul id={id} className={className} style={style}>
      <li><Button className="btn btn-small red lighten-2" onClick={onFakeServerLink}>{`${isFakeServerRunning ? 'Stop' : 'Start'} fake server`}</Button></li>
      <li><Button className="btn btn-small red lighten-2" onClick={simulateThrow}>Simulate a throw</Button></li>
    </ul>
  );
};

MenuLinks.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object,
};

MenuLinks.defaultProps = {
  id: '',
  className: '',
  style: {},
};

export default MenuLinks;
