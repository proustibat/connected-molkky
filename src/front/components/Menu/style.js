export default {
  menu_medium_up: {
    marginTop: '0.4rem',
    display: 'grid',
    gridTemplateColumns: '1fr 1fr',
    gridGap: '1rem',
  },
};
