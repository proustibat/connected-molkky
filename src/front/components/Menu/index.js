import React, { useEffect } from 'react';
import MenuLinks from '@components/Menu/MenuLinks';
import PropTypes from 'prop-types';

const Menu = ({ title }) => {
  useEffect(() => {
    // Mobile menu initialization
    const { M: MaterializeCSS } = window;
    MaterializeCSS && MaterializeCSS.Sidenav.init(document.querySelectorAll('.sidenav'), { edge: 'right' });
  }, []);

  return (
    <div className="navbar-fixed">
      <nav>
        <div className="container nav-wrapper">
          <a href="/molkky/game" className="brand-logo left">{title}</a>
          <a href="/molkky/game" data-target="mobile-menu" className="sidenav-trigger right"><i className="material-icons">menu</i></a>
          <MenuLinks className="right hide-on-med-and-down" />
        </div>
      </nav>
      <MenuLinks id="mobile-menu" className="sidenav right" />
    </div>
  );
};

Menu.propTypes = {
  title: PropTypes.string.isRequired,
};

export default Menu;
