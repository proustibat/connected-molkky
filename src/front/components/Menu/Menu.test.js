import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import MenuLinks from '@components/Menu/MenuLinks';
import Menu from './index';

describe('MenuLinks', () => {
  const { M } = global;
  const title = 'Hello World';
  let querySelectorAllSpy;
  let sidenav;

  beforeAll(() => {
    sidenav = document.createElement('ul');
    sidenav.setAttribute('class', 'sidenav');
    querySelectorAllSpy = jest.spyOn(document, 'querySelectorAll').mockImplementation(() => sidenav);

    delete window.M;
    window.M = { Sidenav: { init: jest.fn() } };
  });

  afterEach(() => {
    querySelectorAllSpy.mockClear();
    window.M.Sidenav.init.mockClear();
  });

  afterAll(() => {
    window.M = M;
    querySelectorAllSpy.mockRestore();
  });

  it('should render the component correctly', () => {
    // Given / When
    const component = shallow(<Menu title={title} />);

    // Then
    expect(component).toHaveLength(1);
    expect(component.find('nav')).toHaveLength(1);
    expect(component.find('nav').find('a')).toHaveLength(2);
    expect(component.find('nav').find('a').at(0).props().children).toBe(title);
    expect(component.find('nav').find('a').at(1).props()).toMatchSnapshot('mobile menu link props');
    expect(component.find(MenuLinks)).toHaveLength(2);
    expect(component.find(MenuLinks).at(1).props()).toMatchSnapshot('mobile menu props');
  });

  it('should initialize the menu on mount', () => {
    // Given / When
    act(() => { mount(<Menu title={title} />); });

    // Then
    expect(global.M.Sidenav.init).toHaveBeenCalledTimes(1);
    expect(querySelectorAllSpy).toHaveBeenCalledTimes(1);
    expect(global.M.Sidenav.init.mock.calls[0][0]).toBe(sidenav);
  });
});
