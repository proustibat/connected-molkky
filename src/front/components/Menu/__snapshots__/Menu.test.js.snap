// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuLinks should render the component correctly: mobile menu link props 1`] = `
Object {
  "children": <i
    className="material-icons"
  >
    menu
  </i>,
  "className": "sidenav-trigger right",
  "data-target": "mobile-menu",
  "href": "/molkky/game",
}
`;

exports[`MenuLinks should render the component correctly: mobile menu props 1`] = `
Object {
  "className": "right hide-on-med-and-down",
  "id": "",
  "style": Object {
    "marginTop": "0.4rem",
  },
}
`;
